spec/function.dd

```diff
@@ -173,124 +173,246 @@ int foo();
+pure int cube(int i)
+{
+    debug writeln("cube() called"); // ok, impure code allowed in debug statement
+    return i * i * i;
+}
+---
 
-        $(UL
-        $(LI read and write the floating point exception flags)
-        $(LI read and write the floating point mode flags, as long as those
-        flags are restored to their initial state upon function entry)
-        $(LI perform impure operations in statements that are in a
-        $(GLINK2 version, ConditionalStatement)
-        controlled by a $(GLINK2 version, DebugCondition).)
-        )
+$(P A pure function can throw exceptions (unless it has the `nothrow` attribute).)
 
-        $(P A pure function can throw exceptions.)
+$(P Note: The rules above do not preclude modification of data obtained through parameters. Unlike
+traditional functional programming languages, D's pure functions allow modification of the caller's
+state through their mutable parameters. This definition of mutable functions is more relaxed than
+the one traditionally employed by pure functional languages because it allows a pure function to
+use state mutation, as long as all state is created internally or reachable through its arguments.
+In particular, a pure function may allocate memory by means of e.g. `new` or `malloc` without these
+being special cases. Example:)
 
 ---
-import std.stdio;
-int x;
-immutable int y;
-const int* pz;
-
-pure int foo(int i,
-             char* p,
-             const char* q,
-             immutable int* s)
-{
-    debug writeln("in foo()"); // ok, impure code allowed in debug statement
-    x = i;   // error, modifying global state
-    i = x;   // error, reading mutable global state
-    i = y;   // ok, reading immutable global state
-    i = *pz; // error, reading const global state
-    return i;
+pure size_t foo(int[] arr) { arr[] += 1; return arr.length; }
+int[] a = [1, 2, 3];
+foo(a);
+assert(a == [2, 3, 4]);
+---
+
+$(P Pure functions fall in exactly one of two categories: $(I weakly pure) and $(I strongly pure).
+There are no dedicated annotations for these; instead, a set of rules applied to the function
+declaration determines its category.)
+
+$(H3 Weak Purity)
+
+$(P A function annotated with `pure` is weakly pure if and only if at least one of the
+following conditions are met:)
+
+$(OL
+$(LI One or more parameter types are incompletely defined. Example:)
+
+---
+struct S; // defined elsewhere
+// These functions are weakly pure:
+pure int fun(S object);
+pure int fun(immutable S object);
+pure int fun(immutable S* pointer);
+---
+
+$(LI One or more parameters allow indirect access$(MDASH)through `ref`, class reference, dynamic
+array, or pointer$(MDASH)to unqualified data or data with qualifiers other than `immutable`.
+Example:)
+
+---
+// These functions are weakly pure:
+pure int fun(Object obj);
+pure int fun(int* x);
+pure int fun(const int[] x);
+struct S
+{
+    const int x;
+}
+pure int fun(S* object);
+---
+
+$(LI The function returns `void`. Example:)
```

Comment:
Why?

Reply:
Currently such functions are already accepted (this diff attempts to clarify the spec without altering it). They don't make sense as strongly pure functions because then they'd be simply not called at all.

An use sample is a deallocation function.

Reply:
Deallocation functions are not a good match for weak purity, because "this will at most modify what is reachable through the parameters" and "this needs to be called to reclaim memory" are not compatible.

Reply:
Note that if we require exceptions and non-termination to be preserved, then strongly pure functions returning `void` may not be elided unless they follow an identical function call.

Reply:
@tgehr I just want to make sure odd cases like that are put in the weak functions category. That means no moving calls around, no elision, and no nasty surprises. Recall that "strongly pure" is the elites, "weakly pure" is the rest. We need to have a good argument to put something in the strongly pure category. Weakly pure is the default.

Reply:
That does not make a lot of sense. Why is a `void` return an "odd case" and returning an empty struct instance is not an "odd case"? The only way this frivolous special casing is acceptable is by noting that we can remove it in the future, but I fear that you might try to double down on it by coupling it with some unrelated feature to save on syntax.

More generally, I am very much against creating categories of pure functions based on the function signature such that the semantics in one category are more relaxed than in another category even though they can call each other. If function inlining can make the set of allowed reorderings smaller, then your semantics is broken.

Reply:
@tgehr I'm just keeping things simple. Going with `void` is simple and obvious. Going with "all data types without state" would work but is more complicated with no increase in expressive power. Simplicity is something to be appreciated.

>More generally, I am very much against creating categories of pure functions based on the function signature such that the semantics in one category are more relaxed than in another category even though they can call each other. If function inlining can make the set of allowed reorderings smaller, then your semantics is broken.

My perception is there's a lot of good experience with distinguishing between weak and strong purity. Initially we only had strong purity, which was very difficult to use. When we added weak purity things got suddenly a ton better. We could of course eliminate them and just go with one definition and let compilers optimize as they find fit. But you'd need a good amount of convincing that these are useless.

So what's your take?

Reply:
> @tgehr I'm just keeping things simple. Going with `void` is simple and obvious. Going with "all data types without state" would work but is more complicated with no increase in expressive power. Simplicity is something to be appreciated.
> ...

Adding an unnecessary special case is not keeping it simple. It is the opposite of that.

We do not even need a way to say: this function is only weakly pure even though it cannot mutate any state. And making it implicit with a `void` return is not even a good way to specify that, even if you do want that feature for some reason. Especially now, when you have allowed `out` parameters on strongly `pure` functions (as well as `pure` factory functions).

> > More generally, I am very much against creating categories of pure functions based on the function signature such that the semantics in one category are more relaxed than in another category even though they can call each other. If function inlining can make the set of allowed reorderings smaller, then your semantics is broken.
> 
> My perception is there's a lot of good experience with distinguishing between weak and strong purity. Initially we only had strong purity, which was very difficult to use. When we added weak purity things got suddenly a ton better. We could of course eliminate them and just go with one definition and let compilers optimize as they find fit. But you'd need a good amount of convincing that these are useless.
> 
> So what's your take?

My take is that your fluffy paragraph fails to address my point while setting up a straw man. (It also wasted a good bit of my time, spent verifying that this is indeed the case.)

Reply:
Sorry for wasting your time. I don't know what I'm doing. 

What I want here is to give people a chance to do systemy things like this:

```
void deallocate(immutable[] array);
```

At some point there's going to be a need for such in a systems programming language. And it should be well defined and work well - no elision, no reordering, just straight call.

Without the special rule, the function above is strongly pure. Right? Then an implementation looks at the rules and figures there's no need to ever call it.

I don't know how to do this another way. Please advise.